### PR TITLE
[e131] Replace std::set with std::vector to reduce flash usage

### PR DESCRIPTION
**Pull request by bdraco**
# What does this implement/fix?

This PR optimizes the E131 component's memory usage by replacing STL container overkill with more appropriate data structures for embedded systems.

## Changes

1. **Replace `std::set` with `std::vector` for `light_effects_`**
   - The `std::set` was overkill for storing 1-4 light effects
   - Red-black tree overhead (~2KB flash) is unnecessary for such small collections
   - Linear search on 1-4 elements is faster than tree traversal anyway
   - Implemented O(1) swap-and-pop removal since iteration order doesn't matter

2. **Remove dead code**
   - Removed `std::map<int, E131Packet> universe_packets_` which was declared but never used anywhere in the codebase

3. **Keep `universe_consumers_` map**
   - This map is justified - it performs reference counting with arbitrary universe numbers
   - O(log n) lookup is appropriate here

### Swap-and-pop removal pattern:
```cpp
// Old: O(n) erase with vector shift
light_effects_.erase(it);

// New: O(1) swap-and-pop (order doesn't matter)
*it = light_effects_.back();
light_effects_.pop_back();
```

This works because `process_()` just iterates and OR's the results - order is irrelevant.

## Types of changes

- [x] Code quality improvements to existing code or addition of tests

**Related issue or feature (if applicable):**

- Part of ongoing embedded systems memory optimization effort

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- N/A - no user-facing changes

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx
- [ ] nRF52840

**Note:** I do not have physical E131 devices to test runtime behavior. However:
- The changes are straightforward container replacements
- Logic has been verified with test harnesses (see attached `e131_optimization_tests.zip`)
- All existing component tests pass compilation
- The swap-and-pop pattern is tested for all edge cases (single element, last element, middle element, etc.)

**Test Archive:** `e131_optimization_tests.zip` contains standalone C++ test programs that verify:
- Swap-and-pop removal works correctly for all edge cases
- E131Component lifecycle (add/remove/process) maintains correct behavior
- Order independence in iteration
- Duplicate detection with std::find
- Graceful handling of edge cases

## Example entry for `config.yaml`:

```yaml
# No config changes - existing E131 configs work as-is
light:
  - platform: neopixelbus
    id: led_strip
    type: GRB
    variant: WS2812
    pin: GPIO2
    num_leds: 60
    effects:
      - e131:
          universe: 1
          channels: RGB
```

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs). (N/A - no user-facing changes)

### CODE REVIEW

**Comment by bdraco** [comment]
[e131_optimization_tests.zip](https://github.com/user-attachments/files/23201753/e131_optimization_tests.zip)

File: esphome/components/e131/e131.cpp [hunk 1]
```diff
@@ -76,29 +78,32 @@ void E131Component::loop() {
 }
 
 void E131Component::add_effect(E131AddressableLightEffect *light_effect) {
-  if (light_effects_.count(light_effect)) {
+  if (std::find(light_effects_.begin(), light_effects_.end(), light_effect) != light_effects_.end()) {
```
Comment (bdraco) [inline esphome/components/e131/e131.cpp:81]: would be slightly less flash to do a linear search but I figured good to not optimize this too much since its not that common and less to review

**Review by jesserockz — APPROVED** [review]
